### PR TITLE
feat: Raise exception if bin-wise modifier data length doesn't match sample data

### LINKED ISSUE #1216: better error message when systematic has wrong number of bins for the channel
# Description

It's good that we can catch samples with wrong number of bins for a channel relatively early. We should catch this earlier too for systematics on the samples. Example here has a systematic with 3 bins for a channel with 4 bins:

```json
{
    "channels": [
        {
            "name": "SR",
            "samples": [
                {
                    "data": [
                        1,
                        2,
                        3,
                        4
                    ],
                    "modifiers": [
                        {
                            "data": null,
                            "name": "SigXsecOverSM",
                            "type": "normfactor"
                        }
                    ],
                    "name": "signal"
                },
                {
                    "data": [
                        2,
                        4,
                        6,
                        8
                    ],
                    "modifiers": [
                        {
                            "data": {
                                "hi_data": [
                                    3,
                                    6,
                                    9
                                ],
                                "lo_data": [
                                    1,
                                    2,
                                    3
                                ]
                            },
                            "name": "syst_wrongbins",
                            "type": "histosys"
                        }
                    ],
                    "name": "WtZ"
                }
            ]
        }
    ],
    "measurements": [
        {
            "config": {
                "parameters": [
                    {
                        "bounds": [
                            [
                                -10.0,
                                10.0
                            ]
                        ],
                        "fixed": true,
                        "inits": [
                            1.0
                        ],
                        "name": "SigXsecOverSM"
                    }
                ],
                "poi": "SigXsecOverSM"
            },
            "name": "NormalMeasurement"
        }
    ],
    "observations": [
        {
            "data": [
                2.0,
                2.0,
                2.0,
                2.0
            ],
            "name": "SR"
        }
    ],
    "version": "1.0.0"
}
```

when running `pyhf cls` on this (or just loading in the workspace), the stacktrace is a bit long and crashes in the combined histosys modifier code (a bit further than we'd like)

```
Traceback (most recent call last):
  File "/Users/kratsg/.pyenv/versions/pyhf-dev/bin/pyhf", line 33, in <module>
    sys.exit(load_entry_point('pyhf', 'console_scripts', 'pyhf')())
  File "/Users/kratsg/.pyenv/versions/3.8.6/envs/pyhf-dev/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kratsg/.pyenv/versions/3.8.6/envs/pyhf-dev/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/kratsg/.pyenv/versions/3.8.6/envs/pyhf-dev/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kratsg/.pyenv/versions/3.8.6/envs/pyhf-dev/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kratsg/.pyenv/versions/3.8.6/envs/pyhf-dev/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/kratsg/pyhf/src/pyhf/cli/infer.py", line 198, in cls
    model = ws.model(
  File "/Users/kratsg/pyhf/src/pyhf/workspace.py", line 412, in model
    return Model(modelspec, poi_name=measurement['config']['poi'], **config_kwargs)
  File "/Users/kratsg/pyhf/src/pyhf/pdf.py", line 561, in __init__
    self.main_model = _MainModel(
  File "/Users/kratsg/pyhf/src/pyhf/pdf.py", line 417, in __init__
    self.modifiers_appliers = {
  File "/Users/kratsg/pyhf/src/pyhf/pdf.py", line 418, in <dictcomp>
    k: c(
  File "/Users/kratsg/pyhf/src/pyhf/modifiers/histosys.py", line 63, in __init__
    if histosys_mods:
  File "/Users/kratsg/pyhf/src/pyhf/interpolators/code4p.py", line 22, in __init__
    self._histogramssets = default_backend.astensor(histogramssets)
  File "/Users/kratsg/pyhf/src/pyhf/tensor/numpy_backend.py", line 214, in astensor
    return np.asarray(tensor_in, dtype=dtype)
  File "/Users/kratsg/.pyenv/versions/3.8.6/envs/pyhf-dev/lib/python3.8/site-packages/numpy/core/_asarray.py", line 83, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: setting an array element with a sequence.
```

The best thing to do is probably say something like "modifier <name> in <sample> has incompatible structure for <channel>" or something similar.

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #1216

Raise an `InvalidModifier` exception if bin-wise modifiers have data that is of a different length than the data for the sample they belong to. The exception attempts to tell the user exactly where the problem is in the spec.

This gets applied to `histosys`, `shapesys`, and `staterror` modifiers if I have this right.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add a check to bin-wise modifiers to ensure that the modifier data
length is equal to the length of the data in the sample containing
the modifier. If this fails, raise a pyhf.exceptions.InvalidModifier
and try to give the user as much information as possible about what
modifier has failed.
   - Apply this to histosys, shapesys, and staterror
* Add tests to test_modifiers for this modifier behavior.
* Apply isort to all changed files
* Remove unused testing code from test_modifiers.py that was missed
for removal in PR #1625
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
I still need to add tests, but @kratsg and @lukasheinrich if you have early feedback on changes they're welcome &mdash; I haven't touched the modifiers in a long time.

File: tests/test_modifiers.py [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
-modifier_pdf_types = ["normal", None, "normal", None, "poisson", "normal"]
+# modifiers_to_test = [
+#     "histosys",
+#     "normfactor",
+#     "normsys",
+#     "shapefactor",
+#     "shapesys",
+#     "staterror",
+# ]
+# modifier_pdf_types = ["normal", None, "normal", None, "poisson", "normal"]
```
Comment (matthewfeickert) [inline tests/test_modifiers.py:12]: These don't seem to be used anywhere, but maybe should get removed in a separate cleaning PR.

Comment (matthewfeickert) [reply tests/test_modifiers.py:12]: Oh they should have gotten removed in PR #1625 but got missed by all of us.

File: tests/test_modifiers.py [hunk 2]
```diff
@@ -174,3 +166,67 @@ def test_shapesys_holes():
...
+            "data": [1, 2, 3],
+        }
+    ]
+    bad_staterror_modifier = [
+        {
+            "name": "staterror_bad",
+            "type": "staterror",
+            "data": [1, 2, 3],
+        }
+    ]
```
Comment (matthewfeickert) [inline tests/test_modifiers.py:223]: If people have recommendations on how to parameterize this test I'm happy to make those changes.

**Comment by matthewfeickert** [comment]
With the spec that @kratsg gave as an example in Issue #1216 as `issue_1216.json` this PR will give

```console
$ pyhf inspect issue_1216.json 
Traceback (most recent call last):
  File "/home/feickert/.pyenv/versions/pyhf-dev-CPU/bin/pyhf", line 33, in <module>
    sys.exit(load_entry_point('pyhf', 'console_scripts', 'pyhf')())
  File "/home/feickert/.pyenv/versions/3.9.6/envs/pyhf-dev-CPU/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/feickert/.pyenv/versions/3.9.6/envs/pyhf-dev-CPU/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/feickert/.pyenv/versions/3.9.6/envs/pyhf-dev-CPU/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/feickert/.pyenv/versions/3.9.6/envs/pyhf-dev-CPU/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/feickert/.pyenv/versions/3.9.6/envs/pyhf-dev-CPU/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/cli/spec.py", line 82, in inspect
    model = ws.model()
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/workspace.py", line 425, in model
    return Model(modelspec, **config_kwargs)
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/pdf.py", line 667, in __init__
    modifiers, _nominal_rates = _nominal_and_modifiers_from_spec(
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/pdf.py", line 148, in _nominal_and_modifiers_from_spec
    finalizd_builder_data[k] = modifiers_builders[k].finalize()
  File "/home/feickert/Code/GitHub/pyhf/src/pyhf/modifiers/histosys.py", line 87, in finalize
    raise InvalidModifier(
pyhf.exceptions.InvalidModifier: The 'WtZ' sample histosys modifier 'syst_wrongbins' has data shape inconsistent with the sample.
WtZ has 'data' of length 4 but syst_wrongbins has 'lo_data' of length 3 and 'hi_data' of length 3.
```

File: tests/test_modifiers.py (hunk 2 above)
Comment (kratsg) [reply tests/test_modifiers.py:223]: make three different workspaces, store as JSON files, and load them up -- and parametrize the paths?

Comment (matthewfeickert) [reply tests/test_modifiers.py:223]: I did this at first, but then it seemed liked I had gone through more work and made it less clear what was actually being tested in each of the 3 situations and so went to this. Thoughts?

Comment (kratsg) [reply tests/test_modifiers.py:223]: hard-coding the spec I think is a better way to go.

Comment (kratsg) [reply tests/test_modifiers.py:223]: But it's probably not as clear.. hrm..

**Review by kratsg — APPROVED** [review]

File: tests/test_modifiers.py (hunk 2 above)
Comment (kratsg) [reply tests/test_modifiers.py:223]: I'll approve for now and unresolve if I convince myself otherwise.

Comment (matthewfeickert) [reply tests/test_modifiers.py:223]: > hard-coding the spec I think is a better way to go.

Okay we can change it then. Or we could make these patch files and patch the spec JSON

**Comment by matthewfeickert** [comment]
@kratsg This is ready for review again now with patch files. I need to run at the moment, but it would be nicer if we could also add newlines at the end of the JSON files.

**Review by kratsg — APPROVED** [review]

**Pull request merged**